ebean-core/src/main/java/io/ebeaninternal/server/persist/BatchControl.java

```diff
@@ -181,15 +179,22 @@ public int executeOrQueue(PersistRequestBean<?> request, boolean batch) throws B
    * Add the request to the batch and return true if we should flush.
    */
   private boolean addToBatch(PersistRequestBean<?> request) {
-    Object alreadyInBatch = persistedBeans.put(request.entityBean(), DUMMY);
+    int depth = transaction.depth();
+    BeanDescriptor<?> desc = request.descriptor();
+    // batching by bean type AND depth
+    String key = desc.rootName() + ":" + depth;
+    
+    String alreadyInBatch = persistedBeans.put(request.entityBean(), key);
     if (alreadyInBatch != null) {
       // special case where the same bean instance has already been
       // added to the batch (doesn't really occur with non-batching
       // as the bean gets changed from dirty to loaded earlier)
-      return false;
+      BatchedBeanHolder beanHolder = getBeanHolder(request, alreadyInBatch);
+      int ordering = depthOrder.orderingFor(depth);
+      return beanHolder.getOrder() > ordering;
```

Comment:
Can you check, if this really makes sense what I've done here ;)

Reply:
Yes, I think it looks ok but I probably need to put a break point here and see some examples so I'll do that.

Reply:
Ah so yes it turns out we can revert this back (we don't need this change) now that we have the other change to "fix the depth" per say.  So I'll revert this part of the change back.